### PR TITLE
Upgrade Express to v4

### PR DESCRIPTION
**Pull request by kevinburkeshyp**
This was accomplished by installing Express 4, attempting to run the tests,
observing things that broke, and then fixing them. In some cases, I leaned on
the work done in https://github.com/balderdashy/sails/pull/3235 to figure out
how to do something. The most comprehensive change is to the router, which is
its own Router object, and no longer a function on an Express `app`.

Adds two new dependencies (which were removed from Express core): cookie-parser
and compression. In some cases we removed the dependency on Express instead of
upgrading - we no longer try to serve favicons, or deal with sessions.

### CODE REVIEW

**Comment by kevinburkeshyp** [comment]
It's also worth noting that Sails core rejected the Express 4 upgrade in favor of calling all of the dependency modules (`compression`, `cookie-parser`, `body-parser`, `serve-favicon`, &c) directly. However, they haven't implemented that yet - they're still leaning on Express 3.

**Comment by benbuckman** [comment]
@kevinburkeshyp Can I make a process request in the future – if there are members of the team who have expressed interest in changing or knowing about particular parts of the system, could you FYI/ping them on PRs that make those changes? For example, I'm very interested in this change (and will review this later today), but would have easily missed the update in Slack if I hadn't happened to see the mention in the #platform channel.

Thanks!

**Comment by kevinburkeshyp** [comment]
Sure - I also tried to ping you in #growth-team

**Comment by benbuckman** [comment]
Oh, my bad – I see that you did ping me there. Thank you!

File: lib/hooks/http/middleware/defaults.js [hunk 1]
```diff
@@ -146,27 +146,11 @@ module.exports = function(sails, app) {
       return res.send(400, bodyParserFailureErrorMsg);
     },
 
-    // By default, the express router middleware is installed towards the end.
-    // This is so that all the built-in core Express/Connect middleware
-    // gets called before matching any explicit routes or implicit shadow routes.
-    router: app.router,
-
     // Add powered-by Sails header
     poweredBy: function xPoweredBy(req, res, next) {
```
Comment (benbuckman) [inline lib/hooks/http/middleware/defaults.js]: Can we turn this off? It's a potential security hole to announce the server software. Express4 (and maybe 3 too) allow it to be disabled with `app.set('x-powered-by', false)` (see [docs](http://expressjs.com/en/4x/api.html)).

Comment (kevinburkeshyp) [inline lib/hooks/http/middleware/defaults.js]: yep totally fine by me, I'll do in separate PR

File: lib/router/bindDefaultHandlers.js [hunk 2]
```diff
@@ -77,7 +81,8 @@ module.exports = function(sails) {
       // Log a message and try to use `res.send` to respond.
       try {
         sails.log.verbose('A request did not match any routes, and no `res.notFound` handler is configured.');
-        res.send(404);
+        res.status(404);
+        res.send('Not found');
```
Comment (benbuckman) [inline lib/router/bindDefaultHandlers.js:85]: FYI, this can also be chained, `res.status(404).send('Not found');`

File: lib/router/index.js [hunk 3]
```diff
@@ -44,7 +44,7 @@ function Router(options) {
   // TODO:
   // instead, use: https://www.npmjs.org/package/path-to-regexp
   // (or: https://www.npmjs.org/package/path-match)
-  this._privateRouter = express();
+  this._privateRouter = express.Router();
```
Comment (benbuckman) [inline lib/router/index.js:47]: _/me wondering if we can assign a whole `Router` to be the "controller" at a wildcard path... opens up all kinds of possibilities like a) versioning, b) phasing out the sails controllers altogether._

File: lib/router/bindDefaultHandlers.js (hunk 2 above)
Comment (kevinburkeshyp) [inline lib/router/bindDefaultHandlers.js:85]: I'm not sure what's going on, but when I tried that, I got `Cannot call method 'send' of undefined`, the response from `res.status` was undefined.

File: lib/router/index.js [hunk 4]
```diff
@@ -226,12 +226,16 @@ Router.prototype.unbind = function(route) {
 
   // Remove route in internal router
   var newRoutes = [];
-  _.each(this._privateRouter.routes[route.method], function(expressRoute) {
-    if (expressRoute.path != route.path) {
+  // Logic taken from https://github.com/balderdashy/sails/pull/3235/files
```
Comment (benbuckman) [inline lib/router/index.js:229]: @kevinburkeshyp do you understand what this is for?

File: package.json [hunk 5]
```diff
@@ -31,10 +31,12 @@
     "captains-log": "0.11.11",
     "colors": "0.6.2",
     "commander": "2.1.0",
+    "compression": "^1.6.1",
+    "cookie-parser": "1.4.1",
     "debug": "2.2.0",
     "ejs": "0.8.8",
     "ejs-locals": "1.0.2",
-    "express": "3.4.3",
+    "express": "4.13.4",
```
Comment (benbuckman) [inline package.json:39]: :clap:

File: test/helpers/router.js [hunk 6]
```diff
@@ -34,11 +34,22 @@ var $Router = {
       this.sails.router.bind.apply(this.sails.router, args);
     });
 
+    it('unbinds the given route', function() {
+      this.sails.router.unbind.should.be.ok;
```
Comment (benbuckman) [inline test/helpers/router.js:38]: is this using `should`? if so, it needs to call the function `.ok()`

Comment (benbuckman) [inline test/helpers/router.js:38]: Hmm, yes it is. The `.should.be.false` and `.should.be.true`s below also need to change then.

File: test/integration/router.specifiedRoutes.test.js [hunk 7]
```diff
@@ -193,36 +193,6 @@ describe('router :: ', function() {
 			});
 		});
 
-		describe('with dynamic url paths specified', function() {
```
Comment (benbuckman) [inline test/integration/router.specifiedRoutes.test.js:196]: Curious why removed?

File: test/unit/router.bind.test.js [hunk 8]
```diff
@@ -1,8 +1,6 @@
 /**
  * Module dependencies
  */
-var supertest = require('supertest');
```
Comment (benbuckman) [inline test/unit/router.bind.test.js:4]: Cool, can this be removed from the dependencies entirely?

File: test/unit/router.bind.test.js [hunk 9]
```diff
@@ -23,10 +19,15 @@ describe('Router.bind', function() {
...
       it('should send expected response (get /foo)', function(done) {
-        supertest(this.sails.router._privateRouter)
-          .get('/foo')
-          .expect(200, 'hello world!')
-          .end(done);
+        this.sails.request({
+          url:'/foo',
+          method: 'get',
+        }, function(err, res, body){
+            if(err) return done(err);
```
Comment (benbuckman) [inline test/unit/router.bind.test.js]: missing space after `if`

File: test/unit/router.bind.test.js [hunk 10]
```diff
@@ -53,58 +54,94 @@ describe('Router.bind', function() {
...
-          .expect(200, 'GET,POST,PUT,DELETE,PATCH')
-          .end(done);
+      // Express 4 (correctly) attempts to send an 'Allow' header via
+      // `res.set('Allow', 'GET, POST')`. This isn't supported by the
+      // mock-res library being used for testing here, so let's skip this
+      // test.
+      it.skip('should send expected response (options /bar)', function(done) {
+        this.sails.request({
+          url:    '/bar',
+          method: 'options'
```
Comment (benbuckman) [inline test/unit/router.bind.test.js:140]: This reminds me, how are we handling CORS in general? (Or do we not need to?)

**Comment by benbuckman** [comment]
@kevinburkeshyp Thank you for doing this. 
I can't tell from this PR what changes are needed in the API to accommodate this – is there a PR on that side too? Let me know if I can help with this.
This opens up tons of opportunities.

File: lib/router/bindDefaultHandlers.js (hunk 2 above)
Comment (benbuckman) [inline lib/router/bindDefaultHandlers.js:85]: Huh, that suggests that Sails is monkey-patching `res.send`...

File: lib/hooks/http/middleware/defaults.js (hunk 1 above)
Comment (kevinburkeshyp) [inline lib/hooks/http/middleware/defaults.js]: https://github.com/Shyp/sails/pull/31

File: lib/router/bindDefaultHandlers.js (hunk 2 above)
Comment (kevinburkeshyp) [inline lib/router/bindDefaultHandlers.js:85]: Maybe this? 

https://github.com/Shyp/sails/blob/upgrade-express/lib/router/res.js#L34-L37

File: lib/router/index.js (hunk 3 above)
Comment (kevinburkeshyp) [inline lib/router/index.js:47]: No idea

File: lib/router/index.js (hunk 4 above)
Comment (kevinburkeshyp) [inline lib/router/index.js:229]: I think this parses routes in `config/routes.js` in different formats and pushes them onto the router stack based on that

Comment (kevinburkeshyp) [inline lib/router/index.js:229]: I don't think we ever unbind routes, though, so I'm not super worried about it

File: test/helpers/router.js (hunk 6 above)
Comment (kevinburkeshyp) [inline test/helpers/router.js:38]: I think this is should 8? so it works?

```
    it('unbinds the given route', function() {
      this.sails.router.unbind = false
      this.sails.router.unbind.should.be.ok;
      var route = {verb: args[0].split(' ')[0], path: args[0].split(' ')[1]};
      this.sails.router.unbind.call(this.sails.router, route);
    });
```

```
  1) sails.router.unbind unbinds the given route:
     AssertionError: expected false to be truthy
    at Object.ok (node_modules/should/lib/should.js:298:10)
    at Context.<anonymous> (test/helpers/router.js:39:41)
```

File: test/integration/router.specifiedRoutes.test.js (hunk 7 above)
Comment (kevinburkeshyp) [inline test/integration/router.specifiedRoutes.test.js:196]: It was asserting things about the internals of express route matching, I believe, specifically the `optional` flags

File: test/unit/router.bind.test.js (hunk 8 above)
Comment (kevinburkeshyp) [inline test/unit/router.bind.test.js:4]: It can, but I'd rather do it in another commit

File: test/unit/router.bind.test.js (hunk 9 above)
Comment (kevinburkeshyp) [inline test/unit/router.bind.test.js]: Yeah, sorry, a lot of these are bad, copypasta from the other PR.

Comment (kevinburkeshyp) [inline test/unit/router.bind.test.js]: 2c85756

File: test/unit/router.bind.test.js (hunk 10 above)
Comment (kevinburkeshyp) [inline test/unit/router.bind.test.js:140]: No idea

**Comment by kevinburkeshyp** [comment]
> I can't tell from this PR what changes are needed in the API to accommodate this – is there a PR on that side too? Let me know if I can help with this.

Not yet, I guess I can run the API tests against this branch and see what passes/fails. I was going to try and get this through first.

**Comment by kevinburkeshyp** [comment]
FYI: https://circleci.com/gh/Shyp/shyp_api/20746
https://github.com/Shyp/shyp_api/compare/express-testing?expand=1

File: package.json [hunk 11]
```diff
@@ -31,10 +31,12 @@
     "captains-log": "0.11.11",
     "colors": "0.6.2",
     "commander": "2.1.0",
+    "compression": "^1.6.1",
```
Comment (caseycrites) [inline package.json]: wildcard.

**Comment by caseycrites** [comment]
One comment, otherwise this looks like :shaved_ice: to me.

File: package.json (hunk 11 above)
Comment (kevinburkeshyp) [inline package.json]: Ah, whoops! Fixed!
